go/ephemeral/erasable_kv_store.go

```diff
@@ -0,0 +1,317 @@
+	return filepath.Join(s.storagePath, key)
+}
+
+func (s *FileErasableKVStore) noiseKey(key string) string {
+	return fmt.Sprintf("%s%s", key, noiseSuffix)
+}
+
+func (s *FileErasableKVStore) getEncryptionKey(ctx context.Context, noiseBytes libkb.NoiseBytes) (fkey [32]byte, err error) {
+	enckey, err := getLocalStorageSecretBoxKey(ctx, s.G())
+	if err != nil {
+		return fkey, err
+	}
+
+	xor, err := libkb.NoiseXOR(enckey, noiseBytes)
+	if err != nil {
+		return fkey, err
+	}
+	copy(fkey[:], xor)
+	return fkey, nil
+}
+
+func (s *FileErasableKVStore) unbox(ctx context.Context, data []byte, noiseBytes libkb.NoiseBytes) (val interface{}, err error) {
+	// Decode encrypted box
+	var boxed boxedData
+	if err := libkb.MPackDecode(data, &boxed); err != nil {
+		return val, err
+	}
+	if boxed.V > cryptoVersion {
+		return val, fmt.Errorf("bad crypto version: %d current: %d", boxed.V,
+			cryptoVersion)
+	}
+	enckey, err := s.getEncryptionKey(ctx, noiseBytes)
+	if err != nil {
+		return val, err
+	}
+	pt, ok := secretbox.Open(nil, boxed.E, &boxed.N, &enckey)
+	if !ok {
+		return val, fmt.Errorf("failed to unbox item")
+	}
+
+	if err = libkb.MPackDecode(pt, val); err != nil {
+		return val, err
+	}
+
+	return val, nil
+}
+
+func (s *FileErasableKVStore) box(ctx context.Context, val interface{}, noiseBytes libkb.NoiseBytes) (data []byte, err error) {
+	data, err = libkb.MPackEncode(val)
+	if err != nil {
+		return data, err
+	}
+
+	enckey, err := s.getEncryptionKey(ctx, noiseBytes)
+	if err != nil {
+		return data, err
+	}
+	var nonce []byte
+	nonce, err = libkb.RandBytes(24)
+	if err != nil {
+		return data, err
+	}
+	var fnonce [24]byte
+	copy(fnonce[:], nonce)
+	sealed := secretbox.Seal(nil, data, &fnonce, &enckey)
+	boxed := boxedData{
+		V: cryptoVersion,
+		E: sealed,
+		N: fnonce,
+	}
+
+	// Encode encrypted box
+	return libkb.MPackEncode(boxed)
+}
+
+func (s *FileErasableKVStore) Put(ctx context.Context, key string, val interface{}) (err error) {
+	s.Lock()
+	defer s.Unlock()
+
+	noiseBytes, err := libkb.MakeNoise()
+	data, err := s.box(ctx, val, noiseBytes)
+	if err != nil {
+		return err
+	}
+	err = s.write(key, data)
+	if err != nil {
+		return err
+	}
+	noiseKey := s.noiseKey(key)
+	return s.write(noiseKey, noiseBytes[:])
+}
+
+func (s *FileErasableKVStore) write(key string, data []byte) (err error) {
+	filepath := s.filepath(key)
+	if err := libkb.MakeParentDirs(s.G().Log, filepath); err != nil {
+		return err
+	}
+
+	tmp, err := ioutil.TempFile(s.storagePath, key)
+	if err != nil {
+		return err
+	}
+	// remove the temp file if it still exists at the end of this function
+	defer libkb.ShredFile(tmp.Name())
+
+	if runtime.GOOS != "windows" {
+		// os.Fchmod not supported on windows
+		if err := tmp.Chmod(libkb.PermFile); err != nil {
+			return err
+		}
+	}
+	if _, err := tmp.Write(data); err != nil {
+		return err
+	}
+	if err := tmp.Close(); err != nil {
+		return err
+	}
+
+	// remove the data if it already exists
+	s.erase(key)
```

Comment:
Pre-existing maybe-bug: I think this step breaks atomicity. It's possible that the rename below fails, in which case we'll have already destroyed the previous value.

On Unix we could solve this by hard linking the old file to a new tmp location, and then shredding it after the rename. On Windows, I think we'd need to somehow call the `ReplaceFile` Win32 function (which Go doesn't expose anywhere as far as I know, so this would require CGO) to take advantage of its `lpBackupFileName` param.

I don't think it's worth doing any of this, since it's a _super_ obscure failure case that we're talking about here. But we might want to leave a fat comment with those details?

Reply:
left a comment and in `secret_store_file.go` too